File: app/src/main/java/me/ykrank/s1next/viewmodel/LoadingViewModel.java
package me.ykrank.s1next.viewmodel;

import android.databinding.BaseObservable;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.IntDef;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

public final class LoadingViewModel extends BaseObservable implements Parcelable {

    public static final Creator<LoadingViewModel> CREATOR = new Creator<LoadingViewModel>() {

        @Override
        public LoadingViewModel createFromParcel(Parcel source) {
            return new LoadingViewModel(source);
        }

        @Override
        public LoadingViewModel[] newArray(int i) {
            return new LoadingViewModel[i];
        }
    };

    public static final int LOADING_FINISH = 0;

    /**
     * We show circular indeterminate {@link android.widget.ProgressBar}
     * for the first time.
     */
    public static final int LOADING_FIRST_TIME = 1;

    public static final int LOADING_SWIPE_REFRESH = 2;

    public static final int LOADING_PULL_UP_TO_REFRESH = 3;
    private int loading;

    public LoadingViewModel() {
    }

    private LoadingViewModel(Parcel source) {
        loading = source.readInt();
    }

    @LoadingDef
    public int getLoading() {
        return loading;
    }

    public void setLoading(@LoadingDef int loading) {
        this.loading = loading;
        notifyChange();
    }

    public boolean isSwipeRefresh() {
        return loading == LOADING_SWIPE_REFRESH;
    }

    public boolean isSwipeRefreshLayoutEnabled() {
        return loading != LOADING_FIRST_TIME && loading != LOADING_PULL_UP_TO_REFRESH;
    }

    public Boolean isLoadingFirstTime() {
        return loading == LOADING_FIRST_TIME;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(loading);
    }

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({
            LOADING_FINISH,
            LOADING_FIRST_TIME,
            LOADING_SWIPE_REFRESH,
            LOADING_PULL_UP_TO_REFRESH
    })
    public @interface LoadingDef {
    }
}

File: library/src/main/java/com/github/ykrank/androidtools/ui/vm/LoadingViewModel.kt
package com.github.ykrank.androidtools.ui.vm

import android.os.Parcel
import androidx.annotation.IntDef
import androidx.databinding.ObservableInt
import androidx.lifecycle.ViewModel

class LoadingViewModel<D> : ViewModel {

    private val loadingObservable = ObservableInt(LOADING_FIRST_TIME)

    @LoadingDef
    var loading: Int
        get() {
            return loadingObservable.get()
        }
        set(value) {
            loadingObservable.set(value)
            loadingObservable.notifyChange()
        }

    var data: D? = null

    constructor()
    private constructor(source: Parcel) {
        loading = source.readInt()
    }

    val isSwipeRefresh: Boolean
        get() = loading == LOADING_SWIPE_REFRESH
    val isSwipeRefreshLayoutEnabled: Boolean
        get() = loading != LOADING_FIRST_TIME && loading != LOADING_PULL_UP_TO_REFRESH
    val isLoadingFirstTime: Boolean
        get() = loading == LOADING_FIRST_TIME

    @Retention(AnnotationRetention.SOURCE)
    @IntDef(value = [LOADING_FINISH, LOADING_FIRST_TIME, LOADING_SWIPE_REFRESH, LOADING_PULL_UP_TO_REFRESH])
    annotation class LoadingDef
    companion object {
        const val LOADING_FINISH = 0

        /**
         * We show circular indeterminate [android.widget.ProgressBar]
         * for the first time.
         */
        const val LOADING_FIRST_TIME = 1
        const val LOADING_SWIPE_REFRESH = 2
        const val LOADING_PULL_UP_TO_REFRESH = 3
    }
}
